src/bin/tripadvisor2mimir.rs

```diff
@@ -37,66 +38,101 @@ struct Settings {
-    let count_ok = Arc::new(AtomicU64::new(0));
-    let count_errors = Arc::new(Mutex::new(HashMap::new()));
-    let pois = {
-        let count_ok = count_ok.clone();
-        let count_errors = count_errors.clone();
-        read_pois(raw_xml, admin_geofinder)
-            .filter_map(move |poi| {
+    let admin_geofinder = build_admin_geofinder(&mimir_es).await;
+
+    // Init Index
+    let index_generator = mimir_es
+        .init_container(&settings.container_tripadvisor)
+        .await
+        .expect("could not create index");
+
+    // Insert POIs
+    let mut indexed_documents = HashSet::new();
+
+    let index_generator = {
+        let raw_xml = read_gzip_file(&settings.tripadvisor.properties).await;
+        let mut count_ok: u64 = 0;
+        let mut count_errors: HashMap<_, u64> = HashMap::new();
+
+        let pois = read_pois(raw_xml, admin_geofinder)
+            .filter_map(|poi| {
                 future::ready(
-                    poi.map_err(|err| {
-                        *count_errors
-                            .lock()
-                            .expect("statistics are not available")
-                            .entry(err)
-                            .or_insert(0) += 1
-                    })
-                    .ok(),
+                    poi.map_err(|err| *count_errors.entry(err).or_insert(0) += 1)
+                        .ok(),
                 )
             })
-            .inspect(move |_| {
-                count_ok.fetch_add(1, Ordering::Relaxed);
-            })
+            .map(|(ta_id, poi)| {
+                indexed_documents.insert(ta_id);
+                count_ok += 1;
+                poi
+            });
+
+        let index_generator = index_generator
+            .insert_documents(pois)
+            .await
+            .expect("could not insert POIs into index");
+
+        info!("Parsed {} POIs", count_ok);
+        info!("Skipped POIs: {:?}", count_errors);
+        index_generator
     };
 
-    // Index POIs
-    mimir_es
-        .generate_index(&settings.container_tripadvisor, pois)
-        .await
-        .expect("error while indexing POIs");
+    // Insert Photos
+    let index_generator = {
+        let raw_xml = read_gzip_file(&settings.tripadvisor.photos).await;
+        let mut count_ok: u64 = 0;
+        let mut count_errors: HashMap<_, u64> = HashMap::new();
+
+        let photos = read_photos(raw_xml)
+            .filter_map(|photos| {
+                future::ready(
+                    photos
+                        .map_err(|err| *count_errors.entry(err).or_insert(0) += 1)
+                        .ok(),
+                )
+            })
+            .filter(|(ta_id, _)| future::ready(indexed_documents.contains(ta_id)))
+            .map(|(ta_id, urls)| {
+                let op = UpdateOperation::Set {
+                    ident: "properties.image".to_string(),
+                    value: urls.join(","),
```

Comment:
Do we need all urls ?

Reply:
Yeah maybe that's overkill, I initially built it this way, then figured out that there was actually only 1 image / property in the feed.

At this point I thought that it could be worth waiting for TA's answer before changing this but as it make take a few day maybe it is better to remove this, and maybe make it in sort that the importer fails if there is a second image someday, it would be a shame if we missed it :cry:

Reply:
I've added a warning, that seems appropriate: https://github.com/Qwant/fafnir/pull/71/commits/20b4d16937bec811db645bea54564f9c1b2c637f#diff-d81682c93bd336cd5c4fe046309675b88c9de2d8aea08f164d769862eb40b5c6R23